controllers/ibu_controller.go

```diff
@@ -306,147 +315,90 @@ func (r *ImageBasedUpgradeReconciler) handleAbortOrFinalize(ctx context.Context,
-	case lcav1alpha1.Stages.Rollback:
-		if !isRollbackAllowed(ibu, isAfterPivot) {
-			utils.SetStatusCondition(&ibu.Status.Conditions,
-				utils.ConditionTypes.RollbackInProgress,
-				utils.ConditionReasons.InvalidTransition,
-				metav1.ConditionFalse,
-				"Upgrade not started or already finalized",
-				ibu.Generation,
-			)
-			return false
-		}
-		utils.SetStatusCondition(&ibu.Status.Conditions,
-			utils.ConditionTypes.UpgradeInProgress,
-			utils.ConditionReasons.Failed,
-			metav1.ConditionFalse,
-			"Rollback requested",
-			ibu.Generation,
-		)
-		utils.SetStatusCondition(&ibu.Status.Conditions,
-			utils.ConditionTypes.UpgradeCompleted,
-			utils.ConditionReasons.Failed,
-			metav1.ConditionFalse,
-			"Rollback requested",
-			ibu.Generation,
-		)
-
-	case lcav1alpha1.Stages.Idle:
-		if isFinalizeAllowed(ibu) {
-			utils.SetStatusCondition(&ibu.Status.Conditions,
-				utils.ConditionTypes.Idle,
-				utils.ConditionReasons.Finalizing,
-				metav1.ConditionFalse,
-				"Finalizing",
-				ibu.Generation,
-			)
-		} else if isAbortAllowed(ibu, isAfterPivot) {
-			utils.SetStatusCondition(&ibu.Status.Conditions,
-				utils.ConditionTypes.Idle,
-				utils.ConditionReasons.Aborting,
-				metav1.ConditionFalse,
-				"Aborting",
-				ibu.Generation,
-			)
-		} else {
-			rollbackCompletedCondition := meta.FindStatusCondition(ibu.Status.Conditions, string(utils.ConditionTypes.RollbackCompleted))
-			idleCondition := meta.FindStatusCondition(ibu.Status.Conditions, string(utils.ConditionTypes.Idle))
-			// Special cases for setting idle when the IBU just got created or after manual cleanup for rollback failure is done
-			if idleCondition == nil {
-				utils.ResetStatusConditions(&ibu.Status.Conditions, ibu.Generation)
-			} else {
-				msg := "Abort or finalize not allowed"
-				if rollbackCompletedCondition != nil && rollbackCompletedCondition.Status == metav1.ConditionFalse {
-					msg = "Transition to Idle not allowed - Rollback failed"
-				} else if isAfterPivot {
-					rollbackInProgressCondition := meta.FindStatusCondition(ibu.Status.Conditions, string(utils.ConditionTypes.RollbackInProgress))
-					if rollbackInProgressCondition != nil && rollbackInProgressCondition.Status == metav1.ConditionTrue {
-						msg = "Transition to Idle not allowed - Rollback is in progress"
-					} else {
-						msg = "Transition to Idle not allowed - Rollback first"
-					}
+	validStages := getValidNextStageList(ibu, isAfterPivot)
+
+	// Check if the stage is in the valid next stage list
+	ok := lo.Contains(validStages, ibu.Spec.Stage)
+	if ok {
+		// The transition is valid, clear any invalid transition conditions if exist and set proper condition
+		switch ibu.Spec.Stage {
+		case lcav1alpha1.Stages.Prep:
+			utils.ClearInvalidTransitionStatusCondition(ibu, lcav1alpha1.Stages.Upgrade)
+			utils.ClearInvalidTransitionStatusCondition(ibu, lcav1alpha1.Stages.Rollback)
+			utils.SetIdleStatusInProgress(ibu, utils.ConditionReasons.InProgress, "In progress")
```

Comment:
This had me confused, admittedly. It threw me that we were setting two conditions to in-progress, and I thought it was a copy/paste error :) ... Maybe a comment here to explicitly note that this is to mark the overall IBU as "in progress", so I don't get confused again next time I see the code

Reply:
Actually, as we now use one logic, I realized that clearing doesn't necessarily have to be handled separatedly for each stage. As long as it transitions to a valid stage, we can clear any existing invalid conditions. It can be a single function in the "if ok" block. It doesn't hurt to check invalid transition in the cases where they are not expected, but it will make this function looking more clear.